### PR TITLE
Deploy to dev environment with Terraform

### PR DESCRIPTION
**Pull request by gunndabad**
# Description

Adds initial Terraform resource definitions and amends the CI pipeline to deploy to the dev environment when PRs are ready for review.

Also adds docs around environment setup, secrets etc.

## Link to Trello Card

https://trello.com/c/PVd5L7Ne/65-terraform-the-api-application

### CODE REVIEW

File: azure/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    "storageAccountName": {
+      "type": "string"
+    },
+    "keyVaultReaderServicePrincipalId": {
+      "type": "string"
+    }
+  },
+  "variables": {
+    "location": "[resourceGroup().location]",
+    "tags": {
```
Comment (saliceti) [inline azure/azuredeploy.json:20]: Do we need to add tags here? I believe CIP automatically propagates tags from the resource group to the resources it contains

File: azure/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+      "Service Line": "Teaching Workforce",
+      "Service Offering": "Database of Qualified Teachers"
+    },
+    "tenantId": "9c7d9dd3-840c-4b3f-818e-552865082e16",
+    "tfstateContainerName": "dqtapi-tfstate",
+    "deliveryTeamGroupId": "ce5793a4-b822-4018-a65d-6c04e96918b2"
+  },
+  "resources": [
+    {
+      "type": "Microsoft.Storage/storageAccounts",
```
Comment (saliceti) [inline azure/azuredeploy.json:35]: Can we enable blob versioning for security?
I think it's via https://docs.microsoft.com/en-us/azure/storage/blobs/versioning-enable?tabs=template

File: azure/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                "backup",
+                "restore"
+              ]
+            }
+          },
+          {
+            "tenantId": "[variables('tenantId')]",
+            "objectId": "[parameters('keyVaultReaderServicePrincipalId')]",
+            "permissions": {
+              "secrets": [
```
Comment (saliceti) [inline azure/azuredeploy.json:83]: Shouldn't we leave only readonly policies?

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -43,18 +47,68 @@ jobs:
     - name: Docker image tag
       id: image
       run: |
-        echo ::set-output name=tag::ghcr.io/$(echo $GITHUB_REPOSITORY | tr '[:upper:]' '[:lower:]'):sha-$(echo $GITHUB_SHA | cut -c -7)
+        echo ::set-output name=tag::$CONTAINER_REGISTRY/$(echo $GITHUB_REPOSITORY | tr '[:upper:]' '[:lower:]'):sha-$(echo $GITHUB_SHA | cut -c -7)
```
Comment (saliceti) [inline .github/workflows/build.yml:50]: Do you mind changing to full SHA as it's what has been used in all other docker images?
such as: https://github.com/DFE-Digital/apply-for-teacher-training/pkgs/container/apply-teacher-training

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -43,18 +47,68 @@ jobs:
...
-        push: ${{ github.ref == 'refs/heads/main' }}
+        push: true
         tags: ${{ steps.image.outputs.tag }}
+
+  validate_terraform:
+    name: Validate Terraform
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/checkout@v1
```
Comment (saliceti) [inline .github/workflows/build.yml:71]: Can we use v2?

File: .github/workflows/actions/deploy-environment/action.yml [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    value: ${{ steps.terraform.outputs.url }}
+
+runs:
+  using: composite
+
+  steps:
+    - name: Get Key Vault name
+      id: keyvault
+      run: |
+        KEY_VAULT_NAME=$(cat $TFVARS | grep -E '^key_vault_name' | cut -d'=' -f2 | tr -d '"' | tr -d ' ')
```
Comment (saliceti) [inline .github/workflows/actions/deploy-environment/action.yml:32]: tfvars.json may also be used to store terraform variables so they can be read with `jq`
example: https://github.com/DFE-Digital/register-trainee-teachers/blob/main/.github/workflows/deploy.yml#L83

File: azure/azuredeploy.json (hunk 1 above)
Comment (gunndabad) [reply azure/azuredeploy.json]: Done. That certainly makes the template simpler.

File: azure/azuredeploy.json (hunk 3 above)
Comment (gunndabad) [reply azure/azuredeploy.json:71]: Amended to have the `get` and `list` permissions only.

File: .github/workflows/build.yml (hunk 5 above)
Comment (gunndabad) [reply .github/workflows/build.yml]: Amended to use v2 throughout

File: .github/workflows/actions/deploy-environment/action.yml (hunk 6 above)
Comment (gunndabad) [reply .github/workflows/actions/deploy-environment/action.yml]: Converted the file to JSON and updated this script to use `jq`.

**Review by saliceti — APPROVED** [review]

**Pull request merged**